### PR TITLE
Update creation of initial environment

### PR DESCRIPTION
**Pull request by trefis**
Merlin and the compiler behaved differently since 4.07 (and ocaml/ocaml#1010) (cf. #1003).
This PR fixes that, and also updates 4.08 to use the new `Load_path` module from the compiler distribution.

While I'm confident about the 4.07 fix, the 4.08 might degrade performances, I'll be monitoring that at janestreet over the next few weeks before merging this PR.

### CODE REVIEW

File: src/ocaml/utils/408/load_path.ml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+module SMap = Misc.StringMap
+
+(* Mapping from basenames to full filenames *)
+type registry = string SMap.t ref
+
+let state = Local_store.new_bindings ()
```
Comment (let-def) [inline src/ocaml/utils/408/load_path.ml]: The `state` value represents a list of dynamically-scoped variables.
With `srefk` create new variables, but at no point in time you redefine these variables.
See https://github.com/ocaml/merlin/blob/master/src/kernel/mocaml.ml#L30 for instance.

The original design assumed that there might be different notion of global scopes, hence the distinction between `Btype.state` and `Env.state`. 
However I think a single `typechecker_state` might be sufficient, and so this notion of binding could be removed.

I will write a PR to simplify this.

File: src/ocaml/typing/408/typemod.ml [hunk 2]
```diff
@@ -189,7 +188,7 @@ let initial_env ~loc ~safe_string ~initially_opened_module
           match after with
           | [] -> None
           | units :: after ->
-              if String.Set.mem m units then
+              if StringSet.mem m units then
```
Comment (let-def) [inline src/ocaml/typing/408/typemod.ml]: In https://github.com/ocaml/merlin/commit/828a79d6121913a61294753555526a7d1739797e (part of the 4.09 support) I propose to keep `String.Set` to stay close to the upstream typechecker style.

With this patch, `Std.String.Set` and `Std.String.Map` are the labelled versions (incompatible with the typechecker code) while `Misc.String.Set` and `Misc.String.Map` are the unlabelled ones.
It works as long as compiler modules open `Misc` module and Merlin modules open `Std` after `Misc`.

File: src/ocaml/utils/408/load_path.ml (hunk 1 above)
Comment (trefis) [reply src/ocaml/utils/408/load_path.ml]: Thanks, I rebased and this should now be fixed.

File: src/ocaml/typing/408/typemod.ml (hunk 2 above)
Comment (trefis) [reply src/ocaml/typing/408/typemod.ml]: 828a79d is a good idea, I extracted it from your PR and pushed it directly on master.
I'll rebase this PR once more (and I guess you'll have to rebase yours)

File: src/ocaml/utils/408/load_path.ml (hunk 1 above)
Comment (trefis) [reply src/ocaml/utils/408/load_path.ml]: Actually, the last commit (which uses `Local_store.Typechecker`) breaks the tests.
I guess there's something I missunderstood.

**Pull request merged**